doc/OnlineDocs/contribution_guide.rst

```diff
@@ -59,6 +59,128 @@ reviewed or merged by the core development team. In addition, any
+
+Developement can then be performed. In order to push your branch to your fork, use
+
+::
+
+   git add <filename>
+   git status  # to check that you have added the correct files
+   git commit -m 'informative commit message to describe changes'
+   git push origin <branch_name>
+
+When your changes are ready for a pull request,
+
+    * visit https://github.com/<username>/pyomo.
+    * Just above the list of files and directories in the repository,
+      you should see a button that says "Branch: master". Click on
+      this button, and choose the correct branch.
+    * Click the "New pull request" button just to the right of the
+      "Branch: <branch_name>" button.
+    * Fill out the pull request template and click the green "Create
+      pull request" button.
+
+Merging the Pyomo master branch into your fork/branch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We recommend updating your fork of Pyomo with changes to Pyomo master
+on a regular basis. This can be done either through GitHub or through
+the command line.
+
+Using GitHub to update your fork
+********************************
+
+    * visit https://github.com/Pyomo/pyomo.
+    * Click on the "New pull request" button just above the list of
+      files and directories.
+    * You will see the title "Compare changes" with some small text
+      below it which says "Compare changes across branches, commits,
+      tags, and more below. If you need to, you can also compare
+      across forks." Click the last part of this: "compare across
+      forks".
+    * You should now see four buttons just below this: "base
+      repository: Pyomo/pyomo", "base: master", "head repository:
+      Pyomo/pyomo", and "compare: master". Click the leftmost button
+      and choose "<username>/Pyomo".
+    * Then click the button which is second to the left, and choose
+      the branch which you want to merge Pyomo master into. The four
+      buttons should now read: "base repository: <username>/pyomo",
+      "base: <branch_name>", "head repository: Pyomo/pyomo", and
+      "compare: master".
+    * You should also now see a pull request template. If you fill out
+      the pull request template and click "Create pull request", this
+      will create a pull request which will update your fork and
+      branch with any changes that have been made to the master branch
+      of Pyomo.
+    * You can then merge the pull request by clicking the green "Merge
+      pull request" button.
+
+Using git commands to update your fork
+**************************************
+
+First, suppose you have set up your remotes with the following:
+
+::
+   
+   git clone https://github.com/<username>/pyomo.git
+   git remote rename origin my_fork
+   git remote add main_fork https://github.com/pyomo/pyomo.git
+
+You can see a list of your remotes with
+
+::
+
+   git remote -v
+
+In this case, you would push to your fork with
+
+::
+
+   git push my_fork <branch_name>
```

Comment:
You technically don't have to list the branch name for remotes other than origin, unless you want the remote branch to have a different name than your local branch name.

Reply:
I personally like the clarity. This explicitly states which branch is being pushed. However, I am still open to making this change. Let's see what other say.

Reply:
Technically, it doesn't say which branch is being pushed - only what branch you are pushing to.  The danger is that you are sitting on the local branch `foo` and then `git push my_fork bar` pushes your local branch `foo` into a different remote branch (I only mention this because I have made this mistake in the past).

Reply:
I agree with John on this one. I never specify the branch name when I push and instead I repeatedly run `git status` while I'm developing to verify the branch I'm on before making any commits or pushing anything. Maybe we could just expand the comment on the `git status` line to double check the active branch (in addition to the staged files).

Reply:
FWIW, I was wrong.  @michaelbynum showed me that when you `git push remote branch`, your local branch `branch` will be pushed up to the remote as the branch `branch` *regardless* of which branch you happen to currently have checked out.  We are pretty sure that what I was doing when I made the mistake was slightly different.